File: src/common/wasm/zig/argwhere.zig
//! WASM argwhere kernels — find flat indices of non-zero elements.
//!
//! argwhere_fill_<T>(a, out, N) → fills `out` with flat indices, returns count.
//!
//! The TS wrapper allocates a worst-case (N u32) output buffer in scratch,
//! then calls fill once and copies out the populated prefix.
//! Scalar loops for all types — the win is avoiding JS overhead.

/// Write flat indices of non-zero f64 elements into `out`. Returns count written.
export fn argwhere_fill_f64(a: [*]const f64, out: [*]u32, N: u32) u32 {
    var count: u32 = 0;
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        if (a[i] != 0) {
            out[count] = i;
            count += 1;
        }
    }
    return count;
}

/// Write flat indices of non-zero f32 elements into `out`. Returns count written.
export fn argwhere_fill_f32(a: [*]const f32, out: [*]u32, N: u32) u32 {
    var count: u32 = 0;
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        if (a[i] != 0) {
            out[count] = i;
            count += 1;
        }
    }
    return count;
}

/// Write flat indices of non-zero i64 elements into `out`. Returns count written.
/// Also handles u64 via reinterpret.
export fn argwhere_fill_i64(a: [*]const i64, out: [*]u32, N: u32) u32 {
    var count: u32 = 0;
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        if (a[i] != 0) {
            out[count] = i;
            count += 1;
        }
    }
    return count;
}

/// Write flat indices of non-zero i32 elements into `out`. Returns count written.
/// Also handles u32 via reinterpret.
export fn argwhere_fill_i32(a: [*]const i32, out: [*]u32, N: u32) u32 {
    var count: u32 = 0;
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        if (a[i] != 0) {
            out[count] = i;
            count += 1;
        }
    }
    return count;
}

/// Write flat indices of non-zero i16 elements into `out`. Returns count written.
/// Also handles u16 via reinterpret.
export fn argwhere_fill_i16(a: [*]const i16, out: [*]u32, N: u32) u32 {
    var count: u32 = 0;
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        if (a[i] != 0) {
            out[count] = i;
            count += 1;
        }
    }
    return count;
}

/// Write flat indices of non-zero i8 elements into `out`. Returns count written.
/// Also handles u8 via reinterpret.
export fn argwhere_fill_i8(a: [*]const i8, out: [*]u32, N: u32) u32 {
    var count: u32 = 0;
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        if (a[i] != 0) {
            out[count] = i;
            count += 1;
        }
    }
    return count;
}

// --- Tests ---

const testing = @import("std").testing;

test "argwhere_fill_f64 basic" {
    const a = [_]f64{ 1.0, 0.0, 3.0, 0.0, 5.0 };
    var out: [5]u32 = undefined;
    const count = argwhere_fill_f64(&a, &out, 5);
    try testing.expectEqual(count, 3);
    try testing.expectEqual(out[0], 0);
    try testing.expectEqual(out[1], 2);
    try testing.expectEqual(out[2], 4);
}

test "argwhere_fill_f64 all zero" {
    const a = [_]f64{ 0.0, 0.0, 0.0 };
    var out: [3]u32 = undefined;
    const count = argwhere_fill_f64(&a, &out, 3);
    try testing.expectEqual(count, 0);
}

test "argwhere_fill_f64 all nonzero" {
    const a = [_]f64{ 1.0, 2.0, 3.0 };
    var out: [3]u32 = undefined;
    const count = argwhere_fill_f64(&a, &out, 3);
    try testing.expectEqual(count, 3);
    try testing.expectEqual(out[0], 0);
    try testing.expectEqual(out[1], 1);
    try testing.expectEqual(out[2], 2);
}

test "argwhere_fill_f64 empty" {
    const a = [_]f64{};
    var out: [0]u32 = undefined;
    const count = argwhere_fill_f64(&a, &out, 0);
    try testing.expectEqual(count, 0);
}

test "argwhere_fill_f32 basic" {
    const a = [_]f32{ 0.0, 1.0, 0.0, 2.0 };
    var out: [4]u32 = undefined;
    const count = argwhere_fill_f32(&a, &out, 4);
    try testing.expectEqual(count, 2);
    try testing.expectEqual(out[0], 1);
    try testing.expectEqual(out[1], 3);
}

test "argwhere_fill_i64 negatives" {
    const a = [_]i64{ -1, 0, -2, 0 };
    var out: [4]u32 = undefined;
    const count = argwhere_fill_i64(&a, &out, 4);
    try testing.expectEqual(count, 2);
    try testing.expectEqual(out[0], 0);
    try testing.expectEqual(out[1], 2);
}

test "argwhere_fill_i32 basic" {
    const a = [_]i32{ 0, 0, 3, 0, 5 };
    var out: [5]u32 = undefined;
    const count = argwhere_fill_i32(&a, &out, 5);
    try testing.expectEqual(count, 2);
    try testing.expectEqual(out[0], 2);
    try testing.expectEqual(out[1], 4);
}

test "argwhere_fill_i16 basic" {
    const a = [_]i16{ 0, 100, 0, 200, 0 };
    var out: [5]u32 = undefined;
    const count = argwhere_fill_i16(&a, &out, 5);
    try testing.expectEqual(count, 2);
    try testing.expectEqual(out[0], 1);
    try testing.expectEqual(out[1], 3);
}

test "argwhere_fill_i8 basic" {
    const a = [_]i8{ 1, 0, -1, 0 };
    var out: [4]u32 = undefined;
    const count = argwhere_fill_i8(&a, &out, 4);
    try testing.expectEqual(count, 2);
    try testing.expectEqual(out[0], 0);
    try testing.expectEqual(out[1], 2);
}

File: src/common/wasm/argwhere.ts
/**
 * WASM-accelerated argwhere — find flat indices of non-zero elements.
 *
 * Returns a Uint32Array of flat indices, or null if WASM can't handle this case.
 * The TS ops layer converts flat indices to multi-indices.
 */

import {
  argwhere_fill_f64,
  argwhere_fill_f32,
  argwhere_fill_i64,
  argwhere_fill_i32,
  argwhere_fill_i16,
  argwhere_fill_i8,
} from './bins/argwhere.wasm';
import {
  resetScratchAllocator,
  resolveInputPtr,
  scratchAlloc,
  getSharedMemory,
  f16InputToScratchF32,
} from './runtime';
import { ArrayStorage } from '../storage';
import { effectiveDType, type DType, TypedArray } from '../dtype';
import { wasmConfig } from './config';

const BASE_THRESHOLD = 32;

type FillFn = (aPtr: number, outPtr: number, N: number) => number;

const fillKernels: Partial<Record<DType, FillFn>> = {
  float64: argwhere_fill_f64,
  float32: argwhere_fill_f32,
  float16: argwhere_fill_f32,
  int64: argwhere_fill_i64,
  uint64: argwhere_fill_i64,
  int32: argwhere_fill_i32,
  uint32: argwhere_fill_i32,
  int16: argwhere_fill_i16,
  uint16: argwhere_fill_i16,
  int8: argwhere_fill_i8,
  uint8: argwhere_fill_i8,
};

type AnyTypedArrayCtor = new (length: number) => TypedArray;
const ctorMap: Partial<Record<DType, AnyTypedArrayCtor>> = {
  float64: Float64Array,
  float32: Float32Array,
  float16: Float32Array,
  int64: BigInt64Array,
  uint64: BigUint64Array,
  int32: Int32Array,
  uint32: Uint32Array,
  int16: Int16Array,
  uint16: Uint16Array,
  int8: Int8Array,
  uint8: Uint8Array,
};

/**
 * WASM-accelerated argwhere for contiguous arrays.
 * Returns a Uint32Array of flat indices of non-zero elements, or null.
 */
export function wasmArgwhereFlat(a: ArrayStorage): Uint32Array | null {
  if (!a.isCContiguous) return null;

  const size = a.size;
  if (size < BASE_THRESHOLD * wasmConfig.thresholdMultiplier) return null;

  const dtype = effectiveDType(a.dtype);
  const kernel = fillKernels[dtype];
  const Ctor = ctorMap[dtype];
  if (!kernel || !Ctor) return null;

  const bpe = (Ctor as unknown as { BYTES_PER_ELEMENT: number }).BYTES_PER_ELEMENT;

  wasmConfig.wasmCallCount++;
  resetScratchAllocator();

  // Resolve input pointer
  let aPtr: number;
  if (dtype === 'float16') {
    aPtr = f16InputToScratchF32(a, size);
  } else {
    aPtr = resolveInputPtr(a.data, a.isWasmBacked, a.wasmPtr, a.offset, size, bpe);
  }

  // Allocate worst-case output (N u32 indices) in scratch
  const outBytes = size * 4; // u32 = 4 bytes
  const outPtr = scratchAlloc(outBytes);

  // Fill indices
  const count = kernel(aPtr, outPtr, size);

  // Copy result out of WASM memory
  const mem = getSharedMemory();
  const wasmView = new Uint32Array(mem.buffer, outPtr, count);
  const result = new Uint32Array(count);
  result.set(wasmView);

  return result;
}
